src/game/logic/object/update/swayclientupdate.cpp

```diff
@@ -0,0 +1,134 @@
+/**
+ * @file
+ *
+ * @author xezon
+ *
+ * @brief Sway client update class. Used for foliage movements.
+ *
+ * @copyright Thyme is free software: you can redistribute it and/or
+ *            modify it under the terms of the GNU General Public License
+ *            as published by the Free Software Foundation, either version
+ *            2 of the License, or (at your option) any later version.
+ *            A full copy of the GNU General Public License can be found in
+ *            LICENSE
+ */
+#include "swayclientupdate.h"
+#include "drawable.h"
+#include "matrix3d.h"
+#include "object.h"
+#include "randomvalue.h"
+#include "scriptengine.h"
+
+SwayClientUpdate::SwayClientUpdate(Thing *thing, const ModuleData *module_data) :
+    ClientUpdateModule(thing, module_data),
+    m_curValue(0.0f),
+    m_curAngle(0.0f),
+    m_curDelta(0.0f),
+    m_curAngleLimit(0.0f),
+    m_leanAngle(0.0f),
+    m_curVersion(-1),
+    m_swaying(true),
+    m_unused(false)
+{
+}
+
+SwayClientUpdate *SwayClientUpdate::Friend_New_Module_Instance(Thing *thing, const ModuleData *module_data)
+{
+    return new SwayClientUpdate(thing, module_data);
+}
+
+void SwayClientUpdate::CRC_Snapshot(Xfer *xfer)
+{
+    ClientUpdateModule::CRC_Snapshot(xfer);
+}
+
+void SwayClientUpdate::Xfer_Snapshot(Xfer *xfer)
+{
+    uint8_t version = 1;
+    xfer->xferVersion(&version, 1);
+    DrawableModule::Xfer_Snapshot(xfer);
+    xfer->xferReal(&m_curValue);
+    xfer->xferReal(&m_curAngle);
+    xfer->xferReal(&m_curDelta);
+    xfer->xferReal(&m_curAngleLimit);
+    xfer->xferReal(&m_leanAngle);
+    xfer->xferShort(&m_curVersion);
+    xfer->xferBool(&m_swaying);
+}
+
+void SwayClientUpdate::Load_Post_Process()
+{
+    ClientUpdateModule::Load_Post_Process();
+    Update_Sway();
+}
+
+NameKeyType SwayClientUpdate::Get_Module_Name_Key() const
+{
+    static NameKeyType s_swayClientUpdateKey = Name_To_Key("SwayClientUpdate");
+    return s_swayClientUpdateKey;
+}
+
+void SwayClientUpdate::Client_Update()
+{
+    if (m_swaying) {
+        Drawable *drawable = Get_Drawable();
+        const BreezeInfo *breeze = g_theScriptEngine->Get_Breeze_Info();
+
+        if (m_curVersion == breeze->version) {
+            if (drawable == nullptr || !drawable->Is_Visible()) {
+                return;
+            }
+        } else {
+            Update_Sway();
+        }
+
+        m_curValue = m_curValue + m_curDelta;
+        if (m_curValue > GAMEMATH_PI2) {
+            m_curValue -= GAMEMATH_PI2;
+        }
+        const float angle = GameMath::Cos(m_curValue);
+        const float new_angle = angle * m_curAngleLimit + m_leanAngle;
+        const float delta_angle = new_angle - m_curAngle;
+        const float rotx = -delta_angle * breeze->sway_direction.x;
+        const float roty = delta_angle * breeze->sway_direction.y;
+
+        Matrix3D new_matrix = drawable->Get_Instance_Matrix();
+        new_matrix.Rotate_X(rotx);
```

Comment:
In_Place_Pre_Rotate_ X
In_Place_Pre_Rotate_Y

Reply:
Tested. Visually looks the same. Trees sway the same way with same speeds. Was this supposed to change something?

Reply:
its possible it won't in place pre and pre sometimes result in same results as normal rotate, but not always, don't know much of matrix maths to tell you how that goes

Reply:
Fixed.